x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/ClampMax.java

```diff
@@ -0,0 +1,206 @@
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        var field = children().get(0);
+        var min = children().get(1);
+        var fieldDataType = field.dataType();
+        TypeResolution resolution = TypeResolutions.isType(
+            field,
+            t -> t.isNumeric() || t == DataType.BOOLEAN || t.isDate() || DataType.isString(t) || t == DataType.IP || t == DataType.VERSION,
+            sourceText(),
+            TypeResolutions.ParamOrdinal.FIRST,
+            fieldDataType.typeName()
+        );
+        if (resolution.unresolved()) {
+            dataType = NULL;
+            return resolution;
+        }
+        if (fieldDataType == NULL) {
+            dataType = NULL;
+            return new TypeResolution("'field' must not be null in clamp()");
+        }
+        resolution = TypeResolutions.isType(
+            min,
+            t -> t.isNumeric() ? fieldDataType.isNumeric() : t.noText() == fieldDataType,
+            sourceText(),
+            TypeResolutions.ParamOrdinal.SECOND,
+            fieldDataType.typeName()
+        );
+        if (resolution.unresolved()) {
+            dataType = NULL;
+            return resolution;
+        }
+        dataType = fieldDataType;
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new ClampMax(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, ClampMax::new, children().get(0), children().get(1));
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        // force datatype initialization
+        var outputType = dataType();
+
+        var max = children().get(1);
+        var maxF = outputType != max.dataType()
+            ? Cast.cast(source(), max.dataType(), outputType, toEvaluator.apply(max))
+            : toEvaluator.apply(max);
+
+        if (dataType == DataType.BOOLEAN) {
+            return new ClampMaxBooleanEvaluator.Factory(source(), toEvaluator.apply(children().get(0)), maxF);
+        }
+        if (dataType == DataType.DOUBLE) {
+            return new ClampMaxDoubleEvaluator.Factory(source(), toEvaluator.apply(children().get(0)), maxF);
+        }
+        if (dataType == DataType.INTEGER) {
+            return new ClampMaxIntegerEvaluator.Factory(source(), toEvaluator.apply(children().get(0)), maxF);
+        }
+        if (dataType == DataType.UNSIGNED_LONG
```

Comment:
Can we convert to element type (ElementType#toElementType) then use it to pick the evaluator?